syntax = "proto3";

package txt2img.v1;

import "google/api/annotations.proto";

option go_package = "github.com/eddiezane/open-source-summit-eu-2022-api-codegen/grpc/client/txt2img/v1;txt2imgv1";

message Image {
  string prompt = 1;
  string id = 2;
  string url = 3;
}

message GenerateImageRequest {
  string prompt = 1;
}

message GenerateImageResponse {
  Image image = 1;
}

message GetImageRequest {
  string id = 1;
}

message GetImageResponse {
  Image image = 1;
}

message ListImagesRequest {
}

message ListImagesResponse {
  repeated Image images = 1;
}

message DeleteImageRequest {
  string id = 1;
}

message DeleteImageResponse {
}

service Txt2ImgService {
  rpc GenerateImage(GenerateImageRequest) returns (GenerateImageResponse) {
    option (google.api.http) = {
      post: "/v1/generate_image"
      body: "*"
    };
  }

  rpc GetImage(GetImageRequest) returns (GetImageResponse) {
    option (google.api.http) = {
      get: "/v1/images/{id}"
    };
  }

  rpc ListImages(ListImagesRequest) returns (ListImagesResponse) {
    option (google.api.http) = {
      get: "/v1/images"
    };
  }

  rpc DeleteImage(DeleteImageRequest) returns (DeleteImageResponse) {
    option (google.api.http) = {
      delete: "/v1/images/{id}"
    };
  }
}
